Abode services, events, lights, cameras, automations, quick actions.

MisterWil opened the pull request:
## Description:

- Updated AbodePy to 0.11.5
- Added config settings, quick action trigger, and capture image services
- Added alarm, panel, and automation events
- Added light and camera platforms
- Added ability to reclassify switches as lights within HA
- Added ability to exclude Abode devices
- Added ability to switch to cloud polling if cloud push events aren't working

**Related issue (if applicable):** N/A

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#3315

## Example entry for `configuration.yaml` (if applicable):
```yaml
abode:
  username: abode_username
  password: abode_password
  name: Abode Alarm System
  polling: False
  exclude:
    - 'ZW:0000000034'
    - 'RF:00000011'
  lights:
    - 'ZW:0000000022'
```

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] Tests have been added to verify that the new code works.

MisterWil committed:
Updated to latest AbodePy version. Added services and events. Added new device types. Added exclude, light, and polling config options.

MisterWil committed:
Disable the event service if polling is enabled.

MisterWil committed:
Merge branch 'dev' into abode

armills commented:
It would be very helpful to break the individual changes into smaller separate PRs. This is a big chunk of review work as it stands now.

arsaboo commented:
@armills The PR admittedly is big, but most of the code is repeated for different platforms (e.g., binary_sensor, cover, etc.). Thus, feedback on one platform can be used across other platforms.

MisterWil commented:
I don't see any easy way to break this PR up into a bunch of smaller PR's, because each "feature" touched each file, which in turn depends on every other change. For example:


- Updated AbodePy to 0.11.5
  - Added additional constructor parameters.
  - Renamed multiple constants
  - Using new constants across existing binary_sensor, lock, switch, and cover platforms
  - Added automation and quick actions to binary_sensor and switch platforms
  - 
- Added config settings, quick action trigger, and capture image services
  - Modified base abode.py to trigger these services, refactored abode.py to separate into new methods
- Added alarm, panel, and automation events
  - This required moving the hass.data[DOMAIN] object away from my abodepy object to a class that holds not only the abodepy object but also data about the loaded devices. This touched every single platform and includes every new platform as well, making this change incompatible with what's currently in dev ENTIRELY.
  - Also utilizes the refactored methods in the base abode.py
- Added light and camera platforms
  - Makes use of new constants available in 0.11.5
- Added ability to reclassify switches as lights within HA
- Added ability to exclude Abode devices
  - Both make use of new abode data class, new config options in base abode.py

Each individual feature depends on some other features changes. I'd love to pull this apart but I just don't see how to do it easily without splitting up this work into hours of additional work and PR's. Not to mention the fact that each change would need to be replicated across all the other PR's. I feel like this will just introduce so many bugs.

In the future I will definitely be submitting smaller PR's... these changes were just required to fully implement the remaining features.

balloob commented on homeassistant/components/abode.py:
```diff
@@ -65,47 +133,149 @@ def setup(hass, config):
+        setting = call.data.get(ATTR_SETTING, None)
+        value = call.data.get(ATTR_VALUE, None)
+
+        if setting and value:
```
No need for this check, you have defined a schema with both required

balloob commented on homeassistant/components/abode.py:
```diff
@@ -65,47 +133,149 @@ def setup(hass, config):
+
+    def capture_image(call):
+        """Capture a new image."""
+        entity_ids = call.data.get(ATTR_ENTITY_ID, None)
```
`None` is the default if no 2nd argument passed in.

balloob commented on homeassistant/components/abode.py:
```diff
@@ -65,47 +133,149 @@ def setup(hass, config):
+    """Home assistant start and stop callbacks."""
+    def startup(event):
+        """Listen for push events."""
+        if not hass.data[DOMAIN].polling:
```
This is not needed since you only listen once and startup event fires once.

balloob commented on homeassistant/components/abode.py:
```diff
@@ -65,47 +133,149 @@ def setup(hass, config):
-    return True
+    def event_callback(event_json, event=None):
+        """Handle an event callback from Abode."""
+        if event:
```
Use a guard clause to save on indentation.

```python
if not event:
    return

data = {
```

balloob replied:
Wait, I'm confused, event can never be `None` because you use partial below to always bind a value.

balloob commented on homeassistant/components/abode.py:
```diff
@@ -65,47 +133,149 @@ def setup(hass, config):
+    for event in events:
+        hass.data[DOMAIN].abode.events.add_event_callback(
+            event,
+            partial(event_callback, event=event))
```
Why not move the event parameter to be the first? that way you can use positional parameters over keyword args

MisterWil replied:
IIRC I placed it second so that the event_callback from the underlying library I wrote would correctly pass the callback data into the first parameter. I'm not super familiar with python partials/parameter details so this was mostly just working off of Google. If you could explain a bit more here or on Discord I'd appreciate it.

balloob replied:
So the idea of partial is that you will auto fill in parameters.

```python
# The method we want to call
def my_callback(first_param, second_param):
    print('first', first_param)
    print('second', second_param)

# This:
def my_wrapper(second_param):
    my_callback('something_for_first', second_param)

# Is equal to:
my_wrapper = partial(my_callback, 'something_for_first')
```

balloob commented on homeassistant/components/abode.py:
```diff
@@ -65,47 +133,149 @@ def setup(hass, config):
     @property
     def should_poll(self):
         """Return the polling state."""
+        if self._data.polling:
```
```python
return self._data.polling
```

balloob commented on homeassistant/components/abode.py:
```diff
@@ -65,47 +133,149 @@ def setup(hass, config):
 
+    def update(self):
+        """Update automation state."""
+        self._device.refresh()
```
Is `refresh()` throttled or will it be called once per update cycle per device?

MisterWil replied:
It is not throttled. I just added back how the polling update worked in the original PR #9030. Should it be done differently?

balloob commented on homeassistant/components/abode.py:
```diff
@@ -124,3 +294,55 @@ def device_state_attributes(self):
+    @property
+    def should_poll(self):
+        """Return the polling state."""
+        if self._data.polling:
```
```python
return self._data.polling
```

balloob commented on homeassistant/components/camera/abode.py:
```diff
@@ -0,0 +1,103 @@
+"""
+This component provides HA camera support for Abode Security System.
```
Next time please keep the PR small. For example, this platform could have been done in a separate PR.

MisterWil replied:
I promise to keep them smaller from now on... I've learned my lesson!

balloob commented on homeassistant/components/camera/abode.py:
```diff
@@ -0,0 +1,103 @@
+            self._device.device_id, self._update_callback
+        )
+
+        if self._event:
```
Confused, there is always an event passed in.

balloob commented on homeassistant/components/camera/abode.py:
```diff
@@ -0,0 +1,103 @@
+class AbodeCamera(AbodeDevice, Camera):
+    """Representation of an Abode camera."""
+
+    def __init__(self, data, device, event=None):
```
Always event passed in, this cannot be `None`. Please remove default value.

balloob commented on homeassistant/components/camera/abode.py:
```diff
@@ -0,0 +1,103 @@
+        """Get a camera image."""
+        self.refresh_image()
+
+        if self._response and self._response.status_code == 200:
```
No need to check for 200 because it wasn't a 200, it will raise during `get_image` and get set to `None`

balloob commented on homeassistant/components/light/abode.py:
```diff
@@ -0,0 +1,96 @@
+"""
+This component provides HA light support for Abode Security System.
```
This one too could have been it's own PR.

balloob commented on homeassistant/components/light/abode.py:
```diff
@@ -0,0 +1,96 @@
+        if device.device_id in data.exclude:
+            continue
+
+        if (device.generic_type == CONST.TYPE_SWITCH and
```
It would make sense to add a helper method on `hass.data[DOMAIN]` that given a device type, will give you a list that takes exclusion into account. That way you are not repeating yourself in every platform.

MartinHjelmare commented on homeassistant/components/binary_sensor/abode.py:
```diff
@@ -17,38 +17,37 @@
     """A binary sensor implementation for Abode device."""
 
-    def __init__(self, controller, device):
+    def __init__(self, data, device):
```
You can remove the init here since you don't implement anything specific for this child class. The parent class init will catch the call arguments.

MartinHjelmare commented on homeassistant/components/binary_sensor/abode.py:
```diff
@@ -58,4 +57,21 @@ def is_on(self):
+class AbodeQuickActionBinarySensor(AbodeAutomation, BinarySensorDevice):
+    """A binary sensor implementation for Abode quick action automations."""
+
+    def __init__(self, data, automation, event):
```
Same as above.

MartinHjelmare commented on homeassistant/components/camera/abode.py:
```diff
@@ -0,0 +1,103 @@
+    @asyncio.coroutine
+    def async_added_to_hass(self):
+        """Subscribe Abode events."""
+        self.hass.async_add_job(
```
```py
yield from super().async_added_to_hass()
```

MisterWil replied:
Should this be added to both? Should this also be added to the same call in components/abode.py?

MartinHjelmare replied:
No. No. The child class is overwriting this method that also exists in the parent class. Instead of duplicating the code that already exist in the parent method, you can call the parent method via `super()`. The second call here in the child method is not the same as in the parent method, so that should stay as is.

MisterWil replied:
Ah, yes, yes, sorry. I was very obviously confused on what the request was.

MartinHjelmare commented on homeassistant/components/cover/abode.py:
```diff
@@ -19,31 +19,34 @@ def setup_platform(hass, config, add_devices, discovery_info=None):
     """Representation of an Abode cover."""
 
-    def __init__(self, controller, device):
+    def __init__(self, data, device):
```
Remove.

MartinHjelmare commented on homeassistant/components/cover/abode.py:
```diff
@@ -19,31 +19,34 @@ def setup_platform(hass, config, add_devices, discovery_info=None):
     @property
     def is_closed(self):
         """Return true if cover is closed, else False."""
         return self._device.is_open is False
```
```py
return not self._device.is_open
```

MartinHjelmare commented on homeassistant/components/light/abode.py:
```diff
@@ -0,0 +1,96 @@
+class AbodeLight(AbodeDevice, Light):
+    """Representation of an Abode light."""
+
+    def __init__(self, data, device):
```
Remove.

MartinHjelmare commented on homeassistant/components/light/abode.py:
```diff
@@ -0,0 +1,96 @@
+        if self._device.is_dimmable and self._device.has_brightness:
+            return self._device.brightness
+
+        return None
```
Not needed.

MartinHjelmare commented on homeassistant/components/light/abode.py:
```diff
@@ -0,0 +1,96 @@
+        if self._device.is_dimmable and self._device.has_color:
+            return self._device.color
+
+        return None
```
Same as above.

MartinHjelmare commented on homeassistant/components/lock/abode.py:
```diff
@@ -19,21 +19,24 @@ def setup_platform(hass, config, add_devices, discovery_info=None):
     """Representation of an Abode lock."""
 
-    def __init__(self, controller, device):
+    def __init__(self, data, device):
```
Remove.

MartinHjelmare commented on homeassistant/components/services.yaml:
```diff
@@ -571,3 +571,32 @@ counter:
+    fields:
+      entity_id:
+        description: Entity id of the quick action to trigger.
+        example: 'binary_sensor.home_quick_action'
```
No blank line at end of file.

MisterWil replied:
Blank line was added -- file is now 603 lines long with the 603rd line blank.

MartinHjelmare commented on homeassistant/components/alarm_control_panel/abode.py:
```diff
@@ -7,7 +7,7 @@
 
 from homeassistant.components.abode import (
-    AbodeDevice, DATA_ABODE, DEFAULT_NAME, CONF_ATTRIBUTION)
+    AbodeDevice, DOMAIN, DEFAULT_NAME, CONF_ATTRIBUTION)
```
You are not allowed to overwrite the `DOMAIN` of a platform. The platform domain is always its base component, eg in this case that's `'alarm_control_panel'`. You can import the abode domain as another name eg:
```py
from homeassistant.components.abode import (..., DOMAIN as ABODE_DOMAIN, ...)
```

MartinHjelmare commented on homeassistant/components/alarm_control_panel/abode.py:
```diff
@@ -22,17 +22,21 @@
 def setup_platform(hass, config, add_devices, discovery_info=None):
     """Set up a sensor for an Abode device."""
-    abode = hass.data[DATA_ABODE]
+    data = hass.data[DOMAIN]
```
Update the abode domain name after the import change above.

MartinHjelmare commented on homeassistant/components/switch/abode.py:
```diff
@@ -18,26 +18,35 @@
     """Representation of an Abode switch."""
 
-    def __init__(self, controller, device):
+    def __init__(self, data, device):
```
Remove.

MartinHjelmare commented on homeassistant/components/switch/abode.py:
```diff
@@ -51,3 +60,24 @@ def turn_off(self, **kwargs):
+class AbodeAutomationSwitch(AbodeAutomation, SwitchDevice):
+    """A switch implementation for Abode automations."""
+
+    def __init__(self, data, automation, event):
```
Remove.

MisterWil commented on homeassistant/components/abode.py:
```diff
@@ -6,57 +6,125 @@
+    ATTR_ENTITY_ID: cv.entity_ids,
+})
+
+CAPTURE_IMAGE_SCHEMA = vol.Schema({
```
Duplicated from above because it is an entirely different schema for an entirely different service request and either one of the schema's could be changed in the future. Just because they are the same now doesn't really mean they should be the same schema does it? Unless this adds some level of absurd overhead I'm not aware of.

MartinHjelmare replied:
The same variable name is used in both the schemas.

MisterWil replied:
...

Oh. lol

MisterWil commented on homeassistant/components/abode.py:
```diff
@@ -65,47 +133,149 @@ def setup(hass, config):
+                ATTR_DATE: event_json.get(ATTR_DATE, ''),
+                ATTR_TIME: event_json.get(ATTR_TIME, ''),
+            }
+            _LOGGER.info(data)
```
Whoops, leftover from testing. :-)

MisterWil committed:
Addressed all CR's

MartinHjelmare commented on homeassistant/components/camera/abode.py:
```diff
@@ -0,0 +1,106 @@
+        """Subscribe Abode events."""
+        yield from super().async_added_to_hass()
+
+        self.hass.async_add_job(
```
You can now remove this call.

MisterWil replied:
Now removed.

MisterWil committed:
Removed duplicated super call.

MartinHjelmare commented on homeassistant/components/alarm_control_panel/abode.py:
```diff
@@ -22,17 +22,21 @@
         """Initialize the alarm control panel."""
-        AbodeDevice.__init__(self, controller, device)
+        super().__init__(data, device)
         self._name = "{0}".format(DEFAULT_NAME)
```
This doesn't seem to be used?

MisterWil committed:
Name config option now used. Removed deprecated DEFAULT_NAME.

MartinHjelmare reviewed (approved):
Looks good. Remaining comments are @balloob's.

balloob reviewed (approved):
This is ok to merge when the comment about `partial` is fixed

MisterWil committed:
Modified partial to move event to first param.

MisterWil commented:
@balloob @MartinHjelmare -- All changes completed as requested. As of this writing cla-bot looks to be lagging but everything else is complete! Thanks, and sorry again for the lengthy review.
